source/extensions/filters/http/lua/lua_filter.cc

```diff
@@ -428,6 +435,129 @@ int StreamHandleWrapper::luaLogCritical(lua_State* state) {
   return 0;
 }
 
+const EVP_MD* StreamHandleWrapper::getDigest(const absl::string_view& name) {
+  std::string hash = absl::AsciiStrToLower(name);
+
+  // Hash Hash algorithms set refers
+  // https://github.com/google/boringssl/blob/ff62b38b4b5a0e7926034b5f93d0c276e55b571d/include/openssl/digest.h
+  if (hash == "md4") {
+    return EVP_md4();
+  } else if (hash == "md5") {
+    return EVP_md5();
+  } else if (hash == "sha1") {
+    return EVP_sha1();
+  } else if (hash == "sha224") {
+    return EVP_sha224();
+  } else if (hash == "sha256") {
+    return EVP_sha256();
+  } else if (hash == "sha384") {
+    return EVP_sha384();
+  } else if (hash == "sha512") {
+    return EVP_sha512();
+  } else if (hash == "md5_sha1") {
+    return EVP_md5_sha1();
+  } else {
+    return nullptr;
+  }
+}
+
+int StreamHandleWrapper::luaVerifySignature(lua_State* state) {
+  // Step 1: get public key
+  auto ptr = lua_touserdata(state, 2);
+  auto key = reinterpret_cast<EVP_PKEY*>(ptr);
+
+  // Step 2: initialize EVP_MD_CTX
+  EVP_MD_CTX* ctx = EVP_MD_CTX_new();
+
+  // Step 3: initialize EVP_MD
+  absl::string_view hash_func = luaL_checkstring(state, 3);
+  const EVP_MD* md = getDigest(hash_func);
+
+  if (md == nullptr) {
+    lua_pushboolean(state, false);
+    std::string err_msg = absl::StrCat(hash_func, " is not supported.");
+    lua_pushlstring(state, err_msg.data(), err_msg.length());
+    EVP_MD_CTX_free(ctx);
+    return 2;
+  }
+
+  // Step 4: initialize EVP_DigestVerify
+  int ok = EVP_DigestVerifyInit(ctx, nullptr, md, nullptr, key);
```

Comment:
I wonder if we can wrap this in some utilities in https://github.com/envoyproxy/envoy/blob/master/source/common/crypto/utility.h. This might then act as an abstraction layer over what BoringSSL vs. OpenSSL provide. @PiotrSikora @bdecoste WDYT?

Reply:
I moved the BoringSSL from Lua bindings to `common/crypto/utility`.